### ISSUE TITLE
Support namespaced attributes

### ISSUE THREAD

**Issue by silenceisgolden** [Polymer/lit-html#423]
Best I can tell, SVG cannot load external content with lit-html. The minimal repro is to use a LitElement based custom element with the template of 

```
<svg>
    <use xlink:href="external.svg#icon-id" />
</svg>
```

with a properly set up external.svg file but I have a full reproduction at https://github.com/silenceisgolden/lit-html-svg-test.

Thanks!

**Comment by Westbrook** [comment]
To work with SVGs and `lit-html` you'll need to use the `svg` template tag a la:

```
import {svg} from 'lit-html/lit-html.js';
...
svg`<svg>
    <use xlink:href="external.svg#mi-edit" />
</svg>`;
```
I've rigged up a demo at https://glitch.com/edit/#!/narrow-gambler?path=index.html:25:16

**Comment by silenceisgolden** [comment]
Changed your demo to be similar to my use case https://glitch.com/edit/#!/helpful-thunder.

As you can see, I need the string in the `xlink:href` to be variable, and something about binding to that attribute causes the example to break.

Thanks for the demo though!

**Comment by Westbrook** [comment]
That is tricky, even when you use the `$=` syntax to set an attribute, it still doesn't display the SVG even though it is setting the attribute in the inspector. https://glitch.com/edit/#!/solid-existence?path=index.html:15:14

You'll also notice that the [`href` attribute](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/href) as outlined in SVG2 does work, but that's not a good solution being Safari has yet to ship support of that.

Wanting to make sure it wasn't a shadow DOM thing I swapped that for standard DOM as well to no luck.

Very strange.

**Comment by silenceisgolden** [comment]
I think this is due to there not being a part that uses `setAttributeNS` which I think is required to dynamically add or update an attribute on `<use>` in SVG1. https://developer.mozilla.org/en-US/docs/Web/API/Element/setAttributeNS for reference. I would expect that a dynamically updated `d` attribute on a path element would have the same bug.

**Comment by jpierson** [comment]
I believe I'm running into this same issue unfortunately in my first test usage of `lit-html` in a project where I'm hoping to adopt it. Are there any known workarounds? @silenceisgolden, I'm I understanding that this may be just the lack of supporting namespaced properties?

**Comment by silenceisgolden** [comment]
@jpierson 

It looks like @justinfagnani or another contributor has not prioritized a fix for this yet so there has not been any discussion of a workaround that I am aware of. However just encapsulating until you have a specific use case where you do not have to use lit-html would be a solution, especially if adding namespaced attribute support would result in more bloat than wanted for the main library.

The solution I will probably go with is to use a vanilla custom element that manages its own shadow DOM template without using lit-html. For a simple element this would be fine but for a more complex element this will be hard to maintain. I would recommend encapsulating at smaller and smaller pieces until you have an element that only needs to deal with namespaced attributes and higher level components can use lit-html. This small element for my use case will look roughly something like:

```
@import { PropertiesMixin } from '@polymer/lib/mixins/properties-mixin.js';

class IconElement extends PropertiesMixin(HTMLElement) {
  static get properties() {
    return {
      iconHref: String
    };
  }

  constructor() {
    super();

    // create shadow root
  }

  connectedCallback() {
    super.connectedCallback();

    // set up shadow root DOM
    // in my use case this will be an <svg> element and a child <use> element
  }

  _propertiesChanged({ iconHref }) {
    // update xlink:href attribute on <use> if iconHref has changed
  }
}
```

**Comment by jpierson** [comment]
I tried an approach to workaround this issue by creating a DOM node and adding it as a template text content parameter such as below as described in under [Getting Started > Writing Templates > Supported Types](https://polymer.github.io/lit-html/guide/writing-templates.html#supported-types) but when rendering I see the following error.

```
Uncaught TypeError: Cannot read property '1' of null
    at new Template (lit-html.js:252)
    at TemplateInstance.defaultTemplateFactory [as _getTemplate] (lit-html.js:97)
    at NodePart._setTemplateResult (lit-html.js:485)
    at NodePart.setValue (lit-html.js:441)
    at NodePart._setIterable (lit-html.js:533)
    at NodePart.setValue (lit-html.js:444)
    at TemplateInstance.update (lit-html.js:587)
    at render (lit-html.js:130)
    at renderBeneficiaries (giwl-enrollment.js:287)
    at renderEnrollment (giwl-enrollment.js:216)
```

Perhaps this is another case where the [documentation is ahead of the release](https://github.com/Polymer/lit-html/issues/432)?

**Comment by ruphin** [comment]
Both of your examples work fine for me, unless there's something I don't understand about your use-case:
https://codepen.io/ruphin/pen/mjgbKG?editors=1010

**Comment by jpierson** [comment]
@ruphin, I believe the issue was with some commented out code sections in the literal HTML. I believe those sections - even though commented out - were still impacting the runtime behavior. Without much change I was able to get the general concept above working so this will be my workaround for the timebeing.

**Comment by ruphin** [comment]
That's unfortunate. As I mentioned before, there should be no workaround necessary. The original example works as intended in my tests.

At the moment lit-html breaks when comments contain escaped values. I'm hoping this will be fixed in a future release.

**Comment by vedtam** [comment]
Same issue here, hope there will be a fix for this.

**Comment by justinfagnani** [comment]
I think we can support namespaced attributes, which should solve this issue. Given the rarity of namespaced attributes though, this is pretty low priority. You can always write a directive for now.

**Comment by mercmobily** [comment]
> At the moment lit-html breaks when comments contain escaped values. I'm hoping this will be fixed in a future release.

is three an issue already open for this? And an easy way to reproduce it?

**Comment by backflip** [comment]
@justinfagnani, is the following one a reasonable approach?

```
const setUseHref = (href) => directive((part) => {
  part.committer.element.setAttributeNS('http://www.w3.org/1999/xlink', 'href', href);
});

<svg>
  <use xlink:href="${setUseHref('something')}" />
</svg>
```

**Comment by ruphin** [comment]
Aside from the new directive syntax (wrap the entire function in the `directive` helper), that directive works.

You can generalise it for use with any attribute:

```js
const xlinkNamespace = directive((value) => (part) => {
  part.committer.element.setAttributeNS('http://www.w3.org/1999/xlink', part.committer.name, href);
});

html`
  <svg>
    <use xlink:href="${xlinkNamespace('something')}" />
  </svg>
`
```

And you can even generalise it further for use with any namespace:

```js
const namespaced = directive((namespace, value) => (part) => {
  part.committer.element.setAttributeNS(namespace, part.committer.name, href);
});
const xlinkNamespace = 'http://www.w3.org/1999/xlink';

html`
  <svg>
    <use xlink:href="${namespaced(xlinkNamespace, 'something')}" />
  </svg>
`
```

The problem with these is that they use some internal implementation details of `AttributePart` which are not part of the external API. That means that whenever the `AttributePart` internals change this directive may break and require some changes. We could consider supporting an "official" `namespaced` directive. Any thoughts on shipping something like the above directive @justinfagnani ? I think the API is nice enough to consider it sufficient support for namespaced attributes.

**Comment by ivands** [comment]
@ruphin your directive works with lit-html but not with lit-element.
Anyone know why?

**Comment by ruphin** [comment]
There was a recent change in how directives are defined that is not backwards compatible. My examples use the new style, and your version of lit-html (as imported by lit-element) probably uses the old style.

Try updating to the latest version to fix your problem.

**Comment by ivands** [comment]
I'm using the latest version of both lit-html & lit-element.
Could it maybe be that the shadow dom is involved?

**Comment by ruphin** [comment]
I can be of more assistance if you create a minimal reproduction of the issue, preferably on codepen or some other online platform.

**Comment by ivands** [comment]
I made a test case on Codepen.
https://codepen.io/anon/pen/gQKgbE?editors=1010
It's not working tho because loading let-element has dependencies that I don't know how to resolve them on Codepen.

I should also mention that the example should be tested on Safari because that's the browser that isn't working for me.

**Comment by justinfagnani** [comment]
@ivands you need to add ?module after the unpkg URLs

**Comment by ivands** [comment]
Ok, the example should work right now.
https://codepen.io/anon/pen/gQKgbE?editors=1010

The only difference is that the href to the SVG is inline because external SVG files can only be loaded from the same domain.
It is important for the example tho because you will see that loading an external SVG file inside a custom element doesn't work in safari.

**Comment by ivands** [comment]
Maybe loading external SVG files into a custom element isn't the way to go.
In chrome the external SVG is loaded each time I use my custom element.
The browser doesn't even cache the SVG. 😭

**Comment by ruphin** [comment]
Your example does not work in lit-element because of Shadow DOM boundaries. You are using ID selectors (`#activity`) in your `use` element, but IDs are scoped inside shadow trees. Lit-element will render your template inside a shadow tree, so it will not have access to the `#activity` node because it is outside that shadow tree.

**Comment by kevinmpowell** [comment]
@ruphin thanks for the directive snippet, it works great, but there's one typo in your generalized example, it should be:
```
const namespaced = directive((namespace, value) => (part) => {
  part.committer.element.setAttributeNS(namespace, part.committer.name, value);
});
const xlinkNamespace = 'http://www.w3.org/1999/xlink';

html`
  <svg>
    <use xlink:href="${namespaced(xlinkNamespace, 'something')}" />
  </svg>
`
```

Yours has `href` in place of `value` in the second example.

**Comment by Acidic9** [comment]
Any updates to this? Would be nice if this could just work with `xlink:href="${imgUrl}"`

**Comment by ruphin** [comment]
The current solution is to use a directive, which works fine for your usecase. An example of such a directive and how to use it can be found up in this thread: https://github.com/Polymer/lit-html/issues/423#issuecomment-432056517

**Comment by justinfagnani** [comment]
I do have a branch with the fix here: https://github.com/Polymer/lit-html/tree/namespaces

I'll update it and make a PR.

**Comment by justinfagnani** [comment]
The PR I had to fix this (#1072) wasn't quite right. Copy and paste comment from there:

> Overall, I don't think the complexity of the fix is worth it. <use xlink:href=""> is the only real use case I'm aware of, and that's deprecated and Safari added support for the plain href attribute in 12.1. I think the recommendation should be to use href or a directive if you need to target older versions of Safari.
